### PR TITLE
Fill holes in ioctl's argument structs

### LINKED ISSUE #169: Fill holes in ioctl's argument structs
See discussion in #150

### PR DESCRIPTION
**Pull request by igaw**
Fixes #169

### CODE REVIEW

File: src/nvme/ioctl.h [hunk 1]
```diff
@@ -384,22 +386,22 @@ int nvme_get_nsid(int fd, __u32 *nsid);
...
+	__u32 timeout;
+	__u32 rsvd12;
+	__u32 *result;
 	enum nvme_identify_cns cns;
 	__u32 nsid;
 	__u16 cntid;
 	__u16 nvmsetid;
 	__u16 domid;
 	__u8 uuidx;
+	__u8 rsvd40[5];
```
Comment (hreinecke) [inline src/nvme/ioctl.h:402]: What's the point of this 'rsvd49' structure?
We really should never do a simple 'memcpy' between this structure and the nvme identify command args, so no point in adding a fudge here.
If you need a reserved field for alignment a simple u8 variable should suffice, no?

File: src/nvme/ioctl.h [hunk 2]
```diff
@@ -428,8 +432,6 @@ static int nvme_identify_cns_nsid(int fd, enum nvme_identify_cns cns,
 		.uuidx = NVME_UUID_NONE,
 		.csi = NVME_CSI_NVM,
 		.data = data,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
```
Comment (hreinecke) [inline src/nvme/ioctl.h:432]: And it's quetionable whether we need to move the initialisation around. Yes, doing so will align with the structure layout, but then we already have a named initializer, so we don't need to rely on the order anymore.
Which was kinda the point of the initial patchset to have a struct as argument.

File: src/nvme/ioctl.h [hunk 3]
```diff
@@ -1137,17 +1137,21 @@ static inline int nvme_zns_identify_ctrl(int fd, struct nvme_zns_id_ctrl *id)
 struct nvme_get_log_args {
 	int args_size;
 	int fd;
+	__u32 timeout;
+	__u32 rsvd12;
```
Comment (hreinecke) [inline src/nvme/ioctl.h:1141]: Argl. No. Why add a reserved field into a perfectly aligned structure?

**Review by hreinecke — CHANGES_REQUESTED** [review]

File: src/nvme/ioctl.h (hunk 1 above)
Comment (igaw) [reply src/nvme/ioctl.h:402]: One thing we can do after introducing the reserved fields is use the pragma attribute packed which makes sure the compiler generates always the same layout.

File: src/nvme/ioctl.h (hunk 2 above)
Comment (igaw) [reply src/nvme/ioctl.h:432]: The idea here is that we have a all common members at the beginning. I was playing with the idea to have a default macro to set these, something like

```
struct nvme_foo_args args = {
    nvme_args_defaults(args, fd)
    .nsid   = 0x1,
    [...],
};
```

Not sure about that though.

File: src/nvme/ioctl.h (hunk 3 above)
Comment (igaw) [reply src/nvme/ioctl.h:1141]: Again the idea is to have match with the common members. So it's for the sake of consistency.

**Comment by igaw** [comment]
BTW, check the discussion in #150 about this topic.

**Comment by igaw** [comment]
Here an alternative version. BTW, I had to update the users of struct nvme_identify_args due to the compiler
issuing this warning:

> ../src/nvme/ioctl.h: In function ‘int nvme_identify_independent_identify_ns(int, __u32, nvme_id_independent_id_ns*)’:
> ../src/nvme/ioctl.h:952:9: error: designator order for field ‘nvme_identify_args::csi’ does not match declaration order in ‘nvme_identify_args’
> 

```
From 8e64efe636cb6febf00b3a20e8657e2469ca7b33 Mon Sep 17 00:00:00 2001
From: Daniel Wagner <dwagner@suse.de>
Date: Thu, 13 Jan 2022 14:19:16 +0100
Subject: [PATCH] ioctl: Rearrange members in nvme_identify_args

Avoid any holes in the struct by rearranging the members. Also add
attribute packed to minimize any ABI changes due compiler
optimizations.

struct nvme_identify_args {
        int                        args_size;            /*     0     4 */
        int                        fd;                   /*     4     4 */
        void *                     data;                 /*     8     8 */
        enum nvme_identify_cns     cns;                  /*    16     4 */
        enum nvme_csi              csi;                  /*    20     4 */
        __u32                      nsid;                 /*    24     4 */
        __u32                      timeout;              /*    28     4 */
        __u32 *                    result;               /*    32     8 */
        __u16                      cntid;                /*    40     2 */
        __u16                      nvmsetid;             /*    42     2 */
        __u16                      domid;                /*    44     2 */
        __u8                       uuidx;                /*    46     1 */

        /* size: 47, cachelines: 1, members: 12 */
        /* last cacheline: 47 bytes */
} __attribute__((__packed__));

Signed-off-by: Daniel Wagner <dwagner@suse.de>
---
 src/nvme/ioctl.h | 139 +++++++++++++++++++++++------------------------
 1 file changed, 69 insertions(+), 70 deletions(-)

diff --git a/src/nvme/ioctl.h b/src/nvme/ioctl.h
index eb16fac423be..2b9dc8a7968f 100644
--- a/src/nvme/ioctl.h
+++ b/src/nvme/ioctl.h
@@ -376,31 +376,31 @@ int nvme_get_nsid(int fd, __u32 *nsid);
 /**
  * nvme_identify_args - Arguments for the NVMe Identify command
  * @fd:		File descriptor of nvme device
+ * @data:	User space destination address to transfer the data
  * @cns:	The Controller or Namespace structure, see @enum nvme_identify_cns
+ * @csi:	Command Set Identifier
  * @nsid:	Namespace identifier, if applicable
  * @domid:	Domain identifier, if applicable
+ * @timeout:	Timeout in ms (0 for default timeout)
+ * @result:	The command completion result from CQE dword0
  * @cntid:	The Controller Identifier, if applicable
  * @nvmsetid:	The NVMe Set ID if CNS is 04h
  * @uuidx:	UUID Index if controller supports this id selection method
- * @csi:	Command Set Identifier
- * @data:	User space destination address to transfer the data
- * @timeout:	Timeout in ms (0 for default timeout)
- * @result:	The command completion result from CQE dword0
  */
 struct nvme_identify_args {
 	int args_size;
 	int fd;
+	void *data;
 	enum nvme_identify_cns cns;
+	enum nvme_csi csi;
 	__u32 nsid;
+	__u32 timeout;
+	__u32 *result;
 	__u16 cntid;
 	__u16 nvmsetid;
 	__u16 domid;
 	__u8 uuidx;
-	enum nvme_csi csi;
-	void *data;
-	__u32 timeout;
-	__u32 *result;
-};
+} __attribute__((packed));
 
 /**
  * nvme_identify() - Send the NVMe Identify command
@@ -420,16 +420,16 @@ static int nvme_identify_cns_nsid(int fd, enum nvme_identify_cns cns,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = data,
 		.cns = cns,
+		.csi = NVME_CSI_NVM,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = data,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -560,16 +560,16 @@ static inline int nvme_identify_ctrl_list(int fd, __u16 cntid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = ctrlist,
 		.cns = NVME_IDENTIFY_CNS_CTRL_LIST,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = cntid,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = ctrlist,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -597,16 +597,16 @@ static inline int nvme_identify_nsid_ctrl_list(int fd, __u32 nsid, __u16 cntid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = ctrlist,
 		.cns = NVME_IDENTIFY_CNS_NS_CTRL_LIST,
+		.csi = NVME_CSI_NVM,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = cntid,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = ctrlist,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -658,16 +658,16 @@ static inline int nvme_identify_nvmset_list(int fd, __u16 nvmsetid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = nvmset,
 		.cns = NVME_IDENTIFY_CNS_NVMSET_LIST,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = nvmsetid,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = nvmset,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -691,16 +691,16 @@ static inline int nvme_identify_primary_ctrl(int fd, __u16 cntid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = cap,
 		.cns = NVME_IDENTIFY_CNS_PRIMARY_CTRL_CAP,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = cntid,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = cap,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -730,16 +730,16 @@ static inline int nvme_identify_secondary_ctrl_list(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = list,
 		.cns = NVME_IDENTIFY_CNS_SECONDARY_CTRL_LIST,
+		.csi = NVME_CSI_NVM,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = cntid,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = list,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -803,16 +803,16 @@ static inline int nvme_identify_ns_csi(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = data,
 		.cns = NVME_IDENTIFY_CNS_CSI_NS,
+		.csi = csi,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = csi,
-		.data = data,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -832,16 +832,16 @@ static inline int nvme_identify_ctrl_csi(int fd, enum nvme_csi csi, void *data)
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = data,
 		.cns = NVME_IDENTIFY_CNS_CSI_CTRL,
+		.csi = csi,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = csi,
-		.data = data,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -870,16 +870,16 @@ static inline int nvme_identify_active_ns_list_csi(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = list,
 		.cns = NVME_IDENTIFY_CNS_NS_ACTIVE_LIST,
+		.csi = csi,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = csi,
-		.data = list,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -908,16 +908,16 @@ static inline int nvme_identify_allocated_ns_list_csi(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = list,
 		.cns = NVME_IDENTIFY_CNS_ALLOCATED_NS_LIST,
+		.csi = csi,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = csi,
-		.data = list,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -939,16 +939,16 @@ static inline int nvme_identify_independent_identify_ns(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = ns,
 		.cns = NVME_IDENTIFY_CNS_CSI_INDEPENDENT_ID_NS,
+		.csi = NVME_CSI_NVM,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = ns,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -989,16 +989,16 @@ static inline int nvme_identify_domain_list(int fd, __u16 domid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = list,
 		.cns = NVME_IDENTIFY_CNS_DOMAIN_LIST,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = domid,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = list,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -1019,16 +1019,16 @@ static inline int nvme_identify_endurance_group_list(int fd, __u16 endgrp_id,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = list,
 		.cns = NVME_IDENTIFY_CNS_ENDURANCE_GROUP_ID,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = endgrp_id,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = list,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -1052,16 +1052,16 @@ static inline int nvme_identify_iocs(int fd, __u16 cntlid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = iocs,
 		.cns = NVME_IDENTIFY_CNS_COMMAND_SET_STRUCTURE,
+		.csi = NVME_CSI_NVM,
 		.nsid = NVME_NSID_NONE,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = cntlid,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
 		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_NVM,
-		.data = iocs,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
@@ -1082,16 +1082,15 @@ static inline int nvme_zns_identify_ns(int fd, __u32 nsid,
 	struct nvme_identify_args args = {
 		.args_size = sizeof(args),
 		.fd = fd,
+		.data = data,
 		.cns = NVME_IDENTIFY_CNS_CSI_NS,
+		.csi = NVME_CSI_ZNS,
 		.nsid = nsid,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
 		.cntid = NVME_CNTLID_NONE,
 		.nvmsetid = NVME_NVMSETID_NONE,
 		.domid = NVME_DOMID_NONE,
-		.uuidx = NVME_UUID_NONE,
-		.csi = NVME_CSI_ZNS,
-		.data = data,
-		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
-		.result = NULL,
 	};
 
 	return nvme_identify(&args);
-- 
2.34.1
```

**Comment by igaw** [comment]
The alternative version is better #171. Let's close this one

**Pull request closed without merging**